README
NAME
    MR::Petrovich::XS - Perl extension for `libpetrovich'

SYNOPSIS
      use MR::Petrovich::XS;

      $petr = MR::Petrovich::XS->instance();
      $petr = MR::Petrovich::XS->instance(config_path => $path_to_yml_config);

      MR::Petrovich::XS->inflect_first_name( $name, $gender, $case )
      MR::Petrovich::XS->inflect_last_name( $name, $gender, $case )
      MR::Petrovich::XS->inflect_middle_name( $name, $gender, $case )

DESCRIPTION
    MR::Petrovich::XS provides an interface to `libpetrovich' library (see
    https://github.com/my-mail-ru/petrovich-c), which allows you inflect
    Russian first, last and middle names. Inflect rules are given from YAML
    file. This file can be found here
    https://github.com/my-mail-ru/petrovich-rules.

CONSTANTS
  KIND
    KIND_FIRST_NAME
    KIND_MIDDLE_NAME
    KIND_LAST_NAME

  GENDER
    GENDER_MALE
    GENDER_FEMALE
    GENDER_ANDROGYNOUS

  CASE
    CASE_NOMINATIVE
    CASE_GENITIVE
    CASE_DATIVE
    CASE_ACCUSATIVE
    CASE_INSTRUMENTAL
    CASE_PREPOSITIONAL

METHODS
    instance(%p)
        Makes instance of MR::Petrovich::XS. `%p' accepts only one parameter
        - `config_path', that defines path to rules config file. Default
        path is `/usr/local/etc/petrovich/rules.yml'

    OBJ->inflect_name($name, $kind, $gender, $case)
        Inflect male or female (see `GENDER_*' constants) `$name', that can
        be first, last or middle part of name (see `KIND_*' constants), to
        choosen case (see `CASE_*' constants). Method returns inflected name
        if result is successful, or original name otherwise.

    OBJ->inflect_first_name($name, $gender, $case)
        Same as `inflect_name', but for first names only.

    OBJ->inflect_middle_name($name, $gender, $case)
        Same as `inflect_name', but for middle names only.

    OBJ->inflect_last_name($name, $gender, $case)
        Same as `inflect_name', but for last names only.

AUTHOR
    Alexander Kazakov, <voland.kot@gmail.com>

COPYRIGHT AND LICENSE
    Copyright (C) 2016 by Alexander Kazakov

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.8 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension MR::Petrovich::XS.

0.01  Wed Jul 13 14:12:49 2016
	- original version; created by h2xs 1.23 with options
		-b 5.8.9 -A -n MR::Petrovich::XS --use-xsloader
0.02  Thu Feb 02 12:38:30 2017
    - fix segmentation fault on empty or undefined strings passed to Petrovich


XS.c
/*
 * This file was generated automatically by ExtUtils::ParseXS version 2.2203 from the
 * contents of XS.xs. Do not edit this file, edit XS.xs instead.
 *
 *	ANY CHANGES MADE HERE WILL BE LOST! 
 *
 */

#line 1 "XS.xs"
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "ppport.h"
#include "petrovich.h"

#define MY_CXT_KEY "MR::Petrovich::XS::_guts" XS_VERSION

typedef struct {
    petr_context_t* petr_ctx;
} my_cxt_t;

START_MY_CXT;

SV * _inflect(const char *data, size_t len, petr_name_kind_t kind, petr_gender_t gender, petr_case_t dest_case) {
    dMY_CXT;

    size_t dest_buf_size = len*4;
    SV *ret = newSV(dest_buf_size);
    SvPOK_on(ret);

    size_t dest_len;
    char *dest = SvPVutf8_nolen(ret);

    int code = petr_inflect(MY_CXT.petr_ctx,data,len,kind,gender,dest_case,dest,dest_buf_size,&dest_len); 
    if(code != 0) { 
        SvREFCNT_dec(ret);
        return &PL_sv_undef;
    }
    
    SvCUR_set(ret, dest_len);
    return ret;
}

#line 46 "XS.c"
#ifndef PERL_UNUSED_VAR
#  define PERL_UNUSED_VAR(var) if (0) var = var
#endif

#ifndef PERL_ARGS_ASSERT_CROAK_XS_USAGE
#define PERL_ARGS_ASSERT_CROAK_XS_USAGE assert(cv); assert(params)

/* prototype to pass -Wmissing-prototypes */
STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params);

STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params)
{
    const GV *const gv = CvGV(cv);

    PERL_ARGS_ASSERT_CROAK_XS_USAGE;

    if (gv) {
        const char *const gvname = GvNAME(gv);
        const HV *const stash = GvSTASH(gv);
        const char *const hvname = stash ? HvNAME(stash) : NULL;

        if (hvname)
            Perl_croak(aTHX_ "Usage: %s::%s(%s)", hvname, gvname, params);
        else
            Perl_croak(aTHX_ "Usage: %s(%s)", gvname, params);
    } else {
        /* Pants. I don't think that it should be possible to get here. */
        Perl_croak(aTHX_ "Usage: CODE(0x%"UVxf")(%s)", PTR2UV(cv), params);
    }
}
#undef  PERL_ARGS_ASSERT_CROAK_XS_USAGE

#ifdef PERL_IMPLICIT_CONTEXT
#define croak_xs_usage(a,b)	S_croak_xs_usage(aTHX_ a,b)
#else
#define croak_xs_usage		S_croak_xs_usage
#endif

#endif

/* NOTE: the prototype of newXSproto() is different in versions of perls,
 * so we define a portable version of newXSproto()
 */
#ifdef newXS_flags
#define newXSproto_portable(name, c_impl, file, proto) newXS_flags(name, c_impl, file, proto, 0)
#else
#define newXSproto_portable(name, c_impl, file, proto) (PL_Sv=(SV*)newXS(name, c_impl, file), sv_setpv(PL_Sv, proto), (CV*)PL_Sv)
#endif /* !defined(newXS_flags) */

#line 98 "XS.c"

XS(XS_MR__Petrovich__XS_init_from_file); /* prototype to pass -Wmissing-prototypes */
XS(XS_MR__Petrovich__XS_init_from_file)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 1)
       croak_xs_usage(cv,  "path");
    {
	const char *	path = (const char *)SvPV_nolen(ST(0));
	int	RETVAL;
	dXSTARG;
#line 48 "XS.xs"
    dMY_CXT;
    RETVAL = petr_init_from_file(path, &MY_CXT.petr_ctx);
#line 117 "XS.c"
	XSprePUSH; PUSHi((IV)RETVAL);
    }
    XSRETURN(1);
}


XS(XS_MR__Petrovich__XS_free_context); /* prototype to pass -Wmissing-prototypes */
XS(XS_MR__Petrovich__XS_free_context)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 0)
       croak_xs_usage(cv,  "");
    {
#line 56 "XS.xs"
    dMY_CXT;
    petr_free_context(MY_CXT.petr_ctx);
#line 138 "XS.c"
    }
    XSRETURN_EMPTY;
}


XS(XS_MR__Petrovich__XS__inflect_name); /* prototype to pass -Wmissing-prototypes */
XS(XS_MR__Petrovich__XS__inflect_name)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 5)
       croak_xs_usage(cv,  "data, len, kind, gender, dest_case");
    {
	const char *	data = (const char *)SvPV_nolen(ST(0));
	size_t	len = (size_t)SvUV(ST(1));
	petr_name_kind_t	kind = (petr_name_kind_t)SvIV(ST(2));
	petr_gender_t	gender = (petr_gender_t)SvIV(ST(3));
	petr_case_t	dest_case = (petr_case_t)SvIV(ST(4));
	SV *	RETVAL;
#line 67 "XS.xs"
    RETVAL = _inflect(data,len,kind,gender,dest_case);
#line 163 "XS.c"
	ST(0) = RETVAL;
	sv_2mortal(ST(0));
    }
    XSRETURN(1);
}


XS(XS_MR__Petrovich__XS__inflect_first_name); /* prototype to pass -Wmissing-prototypes */
XS(XS_MR__Petrovich__XS__inflect_first_name)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 4)
       croak_xs_usage(cv,  "data, len, gender, dest_case");
    {
	const char *	data = (const char *)SvPV_nolen(ST(0));
	size_t	len = (size_t)SvUV(ST(1));
	petr_gender_t	gender = (petr_gender_t)SvIV(ST(2));
	petr_case_t	dest_case = (petr_case_t)SvIV(ST(3));
	SV *	RETVAL;
#line 78 "XS.xs"
    RETVAL = _inflect(data,len,NAME_FIRST,gender,dest_case);
#line 189 "XS.c"
	ST(0) = RETVAL;
	sv_2mortal(ST(0));
    }
    XSRETURN(1);
}


XS(XS_MR__Petrovich__XS__inflect_middle_name); /* prototype to pass -Wmissing-prototypes */
XS(XS_MR__Petrovich__XS__inflect_middle_name)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 4)
       croak_xs_usage(cv,  "data, len, gender, dest_case");
    {
	const char *	data = (const char *)SvPV_nolen(ST(0));
	size_t	len = (size_t)SvUV(ST(1));
	petr_gender_t	gender = (petr_gender_t)SvIV(ST(2));
	petr_case_t	dest_case = (petr_case_t)SvIV(ST(3));
	SV *	RETVAL;
#line 89 "XS.xs"
    RETVAL = _inflect(data,len,NAME_MIDDLE,gender,dest_case);
#line 215 "XS.c"
	ST(0) = RETVAL;
	sv_2mortal(ST(0));
    }
    XSRETURN(1);
}


XS(XS_MR__Petrovich__XS__inflect_last_name); /* prototype to pass -Wmissing-prototypes */
XS(XS_MR__Petrovich__XS__inflect_last_name)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 4)
       croak_xs_usage(cv,  "data, len, gender, dest_case");
    {
	const char *	data = (const char *)SvPV_nolen(ST(0));
	size_t	len = (size_t)SvUV(ST(1));
	petr_gender_t	gender = (petr_gender_t)SvIV(ST(2));
	petr_case_t	dest_case = (petr_case_t)SvIV(ST(3));
	SV *	RETVAL;
#line 100 "XS.xs"
    RETVAL = _inflect(data,len,NAME_LAST,gender,dest_case);
#line 241 "XS.c"
	ST(0) = RETVAL;
	sv_2mortal(ST(0));
    }
    XSRETURN(1);
}

#ifdef __cplusplus
extern "C"
#endif
XS(boot_MR__Petrovich__XS); /* prototype to pass -Wmissing-prototypes */
XS(boot_MR__Petrovich__XS)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
#if (PERL_REVISION == 5 && PERL_VERSION < 9)
    char* file = __FILE__;
#else
    const char* file = __FILE__;
#endif

    PERL_UNUSED_VAR(cv); /* -W */
    PERL_UNUSED_VAR(items); /* -W */
    XS_VERSION_BOOTCHECK ;

        newXS("MR::Petrovich::XS::init_from_file", XS_MR__Petrovich__XS_init_from_file, file);
        newXS("MR::Petrovich::XS::free_context", XS_MR__Petrovich__XS_free_context, file);
        newXS("MR::Petrovich::XS::_inflect_name", XS_MR__Petrovich__XS__inflect_name, file);
        newXS("MR::Petrovich::XS::_inflect_first_name", XS_MR__Petrovich__XS__inflect_first_name, file);
        newXS("MR::Petrovich::XS::_inflect_middle_name", XS_MR__Petrovich__XS__inflect_middle_name, file);
        newXS("MR::Petrovich::XS::_inflect_last_name", XS_MR__Petrovich__XS__inflect_last_name, file);

    /* Initialisation Section */

#line 39 "XS.xs"
{
    MY_CXT_INIT;
    MY_CXT.petr_ctx = NULL;
}

#line 284 "XS.c"

    /* End of Initialisation Section */

#if (PERL_REVISION == 5 && PERL_VERSION >= 9)
  if (PL_unitcheckav)
       call_list(PL_scopestack_ix, PL_unitcheckav);
#endif
    XSRETURN_YES;
}



Makefile.PL
use 5.008008;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'MR::Petrovich::XS',
    VERSION_FROM      => 'lib/MR/Petrovich/XS.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/MR/Petrovich/XS.pm', # retrieve abstract from module
       AUTHOR         => 'Alexander Kazakov <voland.kot@gmail.com>') : ()),
    LIBS              => ['-lpetrovich -lyaml'], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    #OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/MR/Petrovich/XS.pm
=head1 NAME

MR::Petrovich::XS - Perl extension for C<libpetrovich>

=head1 SYNOPSIS

  use MR::Petrovich::XS;

  $petr = MR::Petrovich::XS->instance();
  $petr = MR::Petrovich::XS->instance(config_path => $path_to_yml_config);

  MR::Petrovich::XS->inflect_first_name( $name, $gender, $case )
  MR::Petrovich::XS->inflect_last_name( $name, $gender, $case )
  MR::Petrovich::XS->inflect_middle_name( $name, $gender, $case )

=head1 DESCRIPTION

MR::Petrovich::XS provides an interface to C<libpetrovich> library (see L<https://github.com/my-mail-ru/petrovich-c>), which allows you inflect Russian first, last and middle names.
Inflect rules are given from YAML file. This file can be found here L<https://github.com/my-mail-ru/petrovich-rules>.

=cut

package MR::Petrovich::XS;

use 5.008008;
use strict;
use warnings;
use utf8;

our $VERSION = '0.02';

require XSLoader;
XSLoader::load('MR::Petrovich::XS', $VERSION);

use base qw/Class::Singleton/;
use Carp qw/confess/;

=head1 CONSTANTS

=head2 KIND

=over

=item KIND_FIRST_NAME

=item KIND_MIDDLE_NAME

=item KIND_LAST_NAME

=back

=head2 GENDER

=over

=item GENDER_MALE

=item GENDER_FEMALE

=item GENDER_ANDROGYNOUS

=back

=head2 CASE

=over

=item CASE_NOMINATIVE

=item CASE_GENITIVE

=item CASE_DATIVE

=item CASE_ACCUSATIVE

=item CASE_INSTRUMENTAL

=item CASE_PREPOSITIONAL

=back

=cut

use constant KIND_MAX_IDX   => 2;
use constant CASE_MAX_IDX   => 5;
use constant GENDER_MAX_IDX => 2;

use constant KIND_FIRST_NAME  => 0;
use constant KIND_MIDDLE_NAME => 1;
use constant KIND_LAST_NAME   => 2;

use constant CASE_NOMINATIVE    => 0;
use constant CASE_GENITIVE      => 1;
use constant CASE_DATIVE        => 2;
use constant CASE_ACCUSATIVE    => 3;
use constant CASE_INSTRUMENTAL  => 4;
use constant CASE_PREPOSITIONAL => 5;

use constant GENDER_MALE        => 0;
use constant GENDER_FEMALE      => 1;
use constant GENDER_ANDROGYNOUS => 2;

=head1 METHODS

=over 4

=item instance(%p)

Makes instance of MR::Petrovich::XS. C<%p> accepts only one parameter - C<config_path>, that defines path to rules config file. 
Default path is C</usr/local/etc/petrovich/rules.yml>

=cut

sub _new_instance {
    my ($class, %p) = @_;

    $p{config_path} ||= '/usr/local/etc/petrovich/rules.yml';

    my $self  = bless \%p, $class;
    init_from_file($self->{config_path}); 
    return $self;
}

=item OBJ->inflect_name($name, $kind, $gender, $case)

Inflect male or female (see C<GENDER_*> constants) C<$name>, that can be first, last or middle part of name (see C<KIND_*> constants), to choosen case (see C<CASE_*> constants).
Method returns inflected name if result is successful, or original name otherwise.

=cut
sub inflect_name {
    confess "Wrong arguments count!" unless @_ == 5; 

    my ($self, $name, $kind, $gender, $case) = @_;
    confess "Illegal kind '$kind'!" if $kind < 0 || $kind > KIND_MAX_IDX;
    confess "Illegal gender '$gender'!" if $gender < 0 || $gender > GENDER_MAX_IDX;
    confess "Illegal case '$case'!" if $case < 0 || $case > CASE_MAX_IDX;

    return $name unless $name;

    use bytes;
    my $len = length($name);
    my $inflected_name = _inflect_name($name, $len, $kind, $gender, $case);
    warn "Can't inflect name '$name' (kind:$kind; gender:$gender; case:$case)" if !$inflected_name && $name;
    no bytes;
    return $inflected_name || $name;
}

=item OBJ->inflect_first_name($name, $gender, $case)

Same as C<inflect_name>, but for first names only. 

=cut

sub inflect_first_name {
    my ($self, $name, $gender, $case) = @_;
    return $self->inflect_name($name, KIND_FIRST_NAME, $gender, $case);
}

=item OBJ->inflect_middle_name($name, $gender, $case)

Same as C<inflect_name>, but for middle names only. 

=cut

sub inflect_middle_name {
    my ($self, $name, $gender, $case) = @_;
    return $self->inflect_name($name, KIND_MIDDLE_NAME, $gender, $case);
}

=item OBJ->inflect_last_name($name, $gender, $case)

Same as C<inflect_name>, but for last names only. 

=cut

sub inflect_last_name {
    my ($self, $name, $gender, $case) = @_;
    return $self->inflect_name($name, KIND_LAST_NAME, $gender, $case);
}

sub DESTROY {
    my ($self) = @_;
    free_context();
}

1;

=back

=head1 AUTHOR

Alexander Kazakov, E<lt>voland.kot@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Alexander Kazakov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
